controlplane/segment_allocator.h

```diff
@@ -0,0 +1,308 @@
+				return ErrorResult;
+			}
+			else
+			{
+				// move block from list of free to blocks of this size
+				RemoveFromList(block_index, 0);
+				all_blocks_[block_index].SetSize(size);
+				InsertToList(block_index, size);
+			}
+		}
+
+		OneBlock& block = all_blocks_[size_info.head_block];
+		uint16_t index_in_block = block.Allocate();
+		if (index_in_block == error_in_block_)
+		{
+			errors_internal_++;
+			return ErrorResult;
+		}
+		uint32_t result = size_info.head_block * BlockSize + index_in_block * size + IndexBegin;
+		if (block.free_segments == 0)
+		{
+			// block is busy
+			RemoveFromList(size_info.head_block, size);
+			size_info.busy_blocks++;
+		}
+		size_info.used_segments++;
+
+		free_cells_ -= size;
+		return result;
+	}
+
+	uint32_t Free(uint32_t start, uint16_t size)
+	{
+		if ((size == 0) || (size > MaxBufferSize) || (start >= IndexEnd) || (start < IndexBegin))
+		{
+			errors_external_++;
+			return false;
+		}
+
+		uint32_t block_index = (start - IndexBegin) / BlockSize;
+		OneBlock& block = all_blocks_[block_index];
+		if (block.one_segment_size != size)
+		{
+			errors_external_++;
+			return false;
+		}
+		else if ((start - block_index * BlockSize - IndexBegin) % size != 0)
+		{
+			errors_external_++;
+			return false;
+		}
+
+		bool block_was_busy = (block.free_segments == 0);
+		uint16_t index = (start - block_index * BlockSize - IndexBegin) / size;
+		if (block.Free(index) == error_in_block_)
+		{
+			errors_external_++;
+			return false;
+		}
+
+		OneSizeBlockInfo& size_info = sizes_info_[size];
+		if (block_was_busy)
+		{
+			if ((block.previous != null_block_) || (block.next != null_block_))
+			{
+				errors_internal_++;
+			}
+			InsertToList(block_index, size);
+			size_info.busy_blocks--;
+		}
+		if (block.free_segments == block.total_segments)
+		{
+			RemoveFromList(block_index, size);
+			InsertToList(block_index, 0);
+		}
+
+		size_info.used_segments--;
+		free_cells_ += size;
+		return true;
+	}
+
+	uint32_t Size() const
+	{
+		return free_cells_;
+	}
+
+	const OneSizeBlockInfo* GetBlocksStat() const
+	{
+		return sizes_info_;
+	}
+
+	std::pair<uint64_t, uint64_t> GetErrors() const
+	{
+		return {errors_external_, errors_internal_};
+	}
+
+private:
+	struct OneBlock
+	{
+		// indexes of previous and next blocks in list
+		uint32_t previous;
+		uint32_t next;
+
+		// block structure
+		uint16_t block_size;
+		uint16_t one_segment_size;
+		uint16_t total_segments;
+		uint16_t free_segments;
+
+		// bit masks of segments usage
+		uint64_t group_mask;
+		uint64_t masks[BlockSize / 64];
+
+		void Initialize(uint32_t previous, uint32_t next)
+		{
+			this->previous = previous;
+			this->next = next;
+			block_size = BlockSize;
+			group_mask = 0;
+			memset(masks, 0, sizeof(masks));
+		}
+
+		void SetSize(uint16_t segment_size)
+		{
+			one_segment_size = segment_size;
+			total_segments = block_size / segment_size;
+			free_segments = total_segments;
+		}
+
+		uint16_t Allocate()
+		{
```

Comment:
I don't like the idea of using some magic value as an error in this private method. We can use `std::optional`.
`std::optional<uint16_t> Allocate()`